reference-library/BeforeSwapDelta.md

```diff
@@ -0,0 +1,218 @@
+
+| Param Name | Type    | Description                          |
+|------------|---------|--------------------------------------|
+| deltaSpecified    | int128  | The balance change for the specified token       |
+| deltaUnspecified    | int128  | The balance change for the unspecified token      |
+
+Returns the created `BeforeSwapDelta` value.
+
+This function uses bitwise operations in assembly for gas-efficient packing of the two int128 values:
+
+```solidity
+assembly ("memory-safe") {
+    beforeSwapDelta := or(shl(128, deltaSpecified), and(sub(shl(128, 1), 1), deltaUnspecified))
+}
+```
+
+## Library Functions
+
+### ZERO_DELTA
+
+```solidity
+BeforeSwapDelta public constant ZERO_DELTA = BeforeSwapDelta.wrap(0);
+```
+
+A constant representing a zero delta (no balance changes).
+
+### getSpecifiedDelta
+
+```solidity
+function getSpecifiedDelta(BeforeSwapDelta delta) internal pure returns (int128 deltaSpecified);
+```
+
+Extracts the specified token delta from a `BeforeSwapDelta` value.
+
+| Param Name | Type         | Description                          |
+|------------|--------------|--------------------------------------|
+| delta      | BeforeSwapDelta | The `BeforeSwapDelta` value       |
+
+Returns the extracted specified token delta as an `int128`.
+
+### getUnspecifiedDelta
+
+```solidity
+function getUnspecifiedDelta(BeforeSwapDelta delta) internal pure returns (int128 deltaUnspecified);
+```
+
+Extracts the unspecified token delta from a BeforeSwapDelta value.
+
+| Param Name | Type         | Description                          |
+|------------|--------------|--------------------------------------|
+| delta          | BeforeSwapDelta | The `BeforeSwapDelta`       |
+
+Returns the extracted unspecified token delta as an `int128`.
+
+## Usage in Hooks
+
+When a hook is called during a swap operation, it can perform custom logic and interact with the pool's balances. The `beforeSwap` hook returns a `BeforeSwapDelta` value to indicate any balance changes it introduces. This value can then be forwarded to the `afterSwap` hook, allowing for proper accounting and balance reconciliation.
+The `BeforeSwapDelta` can also be forwarded to the `afterAddLiquidity` and `afterRemoveLiquidity` hooks, enabling consistent balance tracking across various pool operations.
+
+## Usage in PoolManager.sol
+
+Here's how `BeforeSwapDelta` is used in the PoolManager contract:
+
+1. In the `swap` function of the PoolManager contract, the `beforeSwap` hook is called:
+
+```solidity
+function swap(PoolKey memory key, IPoolManager.SwapParams memory params, bytes calldata hookData)
+    external
+    override
+    onlyWhenUnlocked
+    noDelegateCall
+    returns (BalanceDelta swapDelta)
+{
+    // ... (other code)
+
+    BeforeSwapDelta beforeSwapDelta;
+    {
+        int256 amountToSwap;
+        uint24 lpFeeOverride;
+        (amountToSwap, beforeSwapDelta, lpFeeOverride) = key.hooks.beforeSwap(key, params, hookData);
```

Comment:
we need a way to communicate that the amounts in beforeSwapDelta are used against amountSpecified to produce amountToSwap

maybe use an example below?